http/external/src/main/java/org/wildfly/security/http/external/ExternalAuthenticationMechanism.java

```diff
@@ -0,0 +1,107 @@
+ */
+
+package org.wildfly.security.http.external;
+
+import static org.wildfly.security.http.HttpConstants.EXTERNAL_NAME;
+import static org.wildfly.security.http.HttpConstants.FORBIDDEN;
+import static org.wildfly.security.mechanism._private.ElytronMessages.httpExternal;
+
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.sasl.AuthorizeCallback;
+
+import org.wildfly.security.auth.callback.AuthenticationCompleteCallback;
+import org.wildfly.security.http.HttpAuthenticationException;
+import org.wildfly.security.http.HttpServerAuthenticationMechanism;
+import org.wildfly.security.http.HttpServerRequest;
+import org.wildfly.security.mechanism.AuthenticationMechanismException;
+import org.wildfly.security.mechanism._private.MechanismUtil;
+
+/**
+ * The EXTERNAL authentication mechanism.
+ *
+ * @author <a href="mailto:aabdelsa@redhat.com">Ashley Abdel-Sayed</a>
+ */
+public class ExternalAuthenticationMechanism implements HttpServerAuthenticationMechanism {
+
+    private final CallbackHandler callbackHandler;
+
+    ExternalAuthenticationMechanism(CallbackHandler callbackHandler) {
+        this.callbackHandler = callbackHandler;
+    }
+
+    /**
+     * @see org.wildfly.security.http.HttpServerAuthenticationMechanism#getMechanismName()
+     */
+    @Override
+    public String getMechanismName() {
+        return EXTERNAL_NAME;
+    }
+
+    @Override
+    public void evaluateRequest(HttpServerRequest request) throws HttpAuthenticationException {
+
+        String remoteUser = request.getRemoteUser();
+        if (remoteUser == null) {
+            request.noAuthenticationInProgress();
+            return;
+        }
+
+        if (authorize(remoteUser)) {
+            succeed(request);
+        } else {
+            fail(request);
+        }
+
+    }
+
+    private boolean authorize(String username) throws HttpAuthenticationException {
+        AuthorizeCallback authorizeCallback = new AuthorizeCallback(username, username);
+        try {
+            MechanismUtil.handleCallbacks(httpExternal, callbackHandler, authorizeCallback);
+            return authorizeCallback.isAuthorized();
+        } catch (AuthenticationMechanismException e) {
+            throw e.toHttpAuthenticationException();
+        } catch (UnsupportedCallbackException e) {
+            throw httpExternal.mechCallbackHandlerFailedForUnknownReason(e).toHttpAuthenticationException();
+        }
+    }
+
+    private void succeed(HttpServerRequest request) throws HttpAuthenticationException {
+        try {
+            MechanismUtil.handleCallbacks(httpExternal, callbackHandler, AuthenticationCompleteCallback.SUCCEEDED);
+            request.authenticationComplete();
+        } catch (AuthenticationMechanismException e) {
+            throw e.toHttpAuthenticationException();
+        } catch (UnsupportedCallbackException ignored) {
+            // ignored
```

Comment:
Instead of ignoring this, I think we should throw an `HttpAuthenticationException` here.

Reply:
Fixed both, Thanks!